feat(analytics): Add types to modular package

exaby73 opened the pull request:
### Description

See title

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [X] Yes
- My change supports the following platforms;
  - [X] `Android`
  - [X] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [X] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [X] No

dlackty committed:
feat(analytics): implement getSessionId

dlackty committed:
feat(analytics, ios): implement setSessionTimeoutDuration for ios

exaby73 committed:
feat(analytics): Add types to modular package

exaby73 committed:
feat(analytics): Add types to modular package

exaby73 committed:
feat(analytics): Add types to modular package

exaby73 committed:
fix: imports in index.js

mikehardy committed:
test(ios): bump rnfb versions in Podfile.lock

mikehardy committed:
docs(messaging): remove duplicate `title` key

Noticed by @imdkbj
Supercedes #7052

mikehardy committed:
docs(messaging): balance parentheses in text

noticed by @mtalhabaig3
supercedes #7099

Bilal-Abdeen committed:
docs(analytics, android): how to see events in DebugView

mikehardy committed:
docs(messaging): add comments to sample code

Noticed by @mtalhabaig2
Supercedes #7103
Fixes #7073

skam22 committed:
docs(app-check, ios): update installation steps

Additional modifications to AppDelegate.mm are required in order to load the RNFBAppCheckModule.  Without initializing the AppCheckModule before calling [FIRApp configure], the setup of a custom provider will fail.

aadito123 committed:
fix(app-check, types): ReactNativeFirebaseAppCheckProvider.configure returns void not Promise<void>

zhiqingchen committed:
docs(auth, phone): add missing useEffect import in sample code

Salakar committed:
chore(release): release packages

mikehardy committed:
test(remote-config): remove v8-compat test duplication

the modular tests include a v8 compatibility test section, which
was otherwise duplicate of the main test file

move the modular test file into the main test file's place

exaby73 committed:
feat(remote-config): realtime config updates

mikehardy committed:
test(deps): update functions node dependencies

galezra committed:
docs(auth): fixed typo in facebook example code

mikehardy committed:
test(deps): bump versions in podfile lock

mikehardy committed:
test(android): pin the detox websocket so forwarding is stable

This is speculative and unique to a dev environment situation I have
where the emulator is on a different machine and it is more convenient
for me to forward the ports required if they are stable instead of
ephemeral. I don't *think* this will affect anyone else negatively as
we are not running multiple detox instances on the same machine that
I know of, but this may be removed without discussion if it causes
harm to anyone anytime

mikehardy committed:
feat(app, sdks)!: ios-sdk 10.8.0 requires Xcode 13.3+; android-sdk 31.5.0

BREAKING CHANGE: firebase-ios-sdk 10.8.0 and higher require Xcode 13.3+,
which transitively requires macOS 12.0+. You must update your CI build environments
to meet these minimums as well as your development environments - if you have older
hardware that still works but cannot be upgraded normally, you may like:
https://dortania.github.io/OpenCore-Legacy-Patcher/

mikehardy committed:
feat(app, sdk)!: android-sdk v32 - app-check safetyNet provider is removed

Note that forcing server retrieval of firestore named queries against loaded
firestore data bundles may fail on android with this SDK version. It appears
results will always come from the local cache.

BREAKING CHANGE: the "safetyNet" provider for App Check has been removed
from the underlying firebase-android-sdk and we have removed it here. You
should upgrade to the "playIntegrity" provider for App Check

mikehardy committed:
feat(app, sdk)!: ios-sdk 10.10.0, requires Xcode 14.1+ / macOS 12.5+

BREAKING CHANGE: this version of the underlying firebase-ios-sdk has
a minimum Xcode requirement of 14.1 which transitively implies a macOS
minimum version of 12.5

mikehardy committed:
test(ios): increase workflow time allowance

- pod install can take as much as 16 minutes now
- compile + pod install on new SDKs can take more than 60 minutes combined

mikehardy committed:
test(ci): nick-invision --> nick-fields / action repo moved

mikehardy committed:
test(ci): ccache-action has a floating v1 tag now, use it

mikehardy committed:
test(deps): update webfactory/ssh-agent action

Bilal-Abdeen committed:
docs(functions): show how to use region-specific function (#6993)

Co-authored-by: Mike Hardy <github@mikehardy.net>

ashuvssut committed:
docs(auth): update links to full auth error messages list (#7123)

Co-authored-by: Mike Hardy <github@mikehardy.net>

mikehardy committed:
docs(expo): update link to vscode-expo extension

Supercedes #7135

mikehardy committed:
test(ci): allow more time for android tests

mikehardy committed:
test(ci): switch to macos-13 runners for e2e

- allows access to latest Xcode (14.3 installed currently on runners)

falleco committed:
fix(firestore, types): add types for Filter constraints on Queries (#7124)

ibobo committed:
fix(firestore): Allow queries with combined in and array-contains-any (#7142)

billykwok committed:
feat(auth, android): add forceRecaptchaFlowForTesting API (#7148)

Co-authored-by: Mike Hardy <github@mikehardy.net>

mikehardy committed:
docs(auth, multi-factor): correct conditional in code example

Supercedes #7156 - credit to @katayama8000

Salakar committed:
chore(release): release packages

dependabot[bot] committed:
build(deps): bump actions/stale from 6 to 8

Bumps [actions/stale](https://github.com/actions/stale) from 6 to 8.
- [Release notes](https://github.com/actions/stale/releases)
- [Changelog](https://github.com/actions/stale/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/stale/compare/v6...v8)

---
updated-dependencies:
- dependency-name: actions/stale
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

exaby73 committed:
feat(installations): Firebase JS SDK V9 modular API (#7095)

mikehardy commented on packages/analytics/lib/modular/index.js:
```diff
@@ -569,6 +619,7 @@ export function isSupported() {
   // always return "true" for now until Web implementation. Web only.
   return Promise.resolve(true);
 }
+
```
There are unnecessary spacing changes all over this diff, somewhere between 100 and 200 lines show up as visible changes in the PR diff viewer without representing any functional changes. This makes diffs harder to read and should be avoided

mikehardy commented on packages/analytics/lib/index.d.ts:
```diff
@@ -1675,6 +1705,42 @@ export namespace FirebaseAnalyticsTypes {
+   */
+  export declare type CustomEventName<T> = T extends EventNameString ? never : T;
+  /**
+   * Type for standard Google Analytics event names. logEvent also accepts any custom string and interprets it as a custom event name.
```
Is there a URL we could place here to serve as a reference for the standard names (both for help maintaining in the future - these names change over time, even if slowly and we need to stay in sync) as well as for folks in the community (who may provide PRs in the future to update the event names, this has happened in the past)

exaby73 replied:
This type comes from [the docs](https://firebase.google.com/docs/reference/js/analytics.md#customeventname). The above doc comment is all that is mentioned in [the docs](https://firebase.google.com/docs/reference/js/analytics.md#customeventname)

mikehardy replied:
Hey @exaby73 - I wasn't clear, sorry. There are a huge number of strings here starting at line 1717, (example 'add_payment_info').

Where did they come from? I'd like a link to the list of things in EventNameString so that future maintainers will have a reference when trying to keep them in sync.

The question to ask your future self one year from now is: "Are these names still the valid names?", and the action to take is the action your future self wished your past self would take: that is, leave your future self an easy way to complete the inevitable task, while you are here working on it now...

mikehardy replied:
@exaby73 no response here. I need a response here - I'm expecting to see some link to some source of truth for this list so it may be compared now and in the future for correctness / freshness

exaby73 replied:
https://github.com/invertase/react-native-firebase/pull/7109/commits/6d0f422cd0f017ba82fd97a11702d949fcaf48f1

mikehardy replied:
Yes! Thank you

mikehardy commented on packages/analytics/lib/index.js:
```diff
@@ -84,7 +84,7 @@ export {
   setDefaultEventParameters,
   initiateOnDeviceConversionMeasurementWithEmailAddress,
-} from '../modular/index';
+} from './modular/index';
```
amazing that worked before! How did that work in fact, without testing indicating some sort of failure 🤔

exaby73 replied:
I moved the file from outside `lib` to inside it to follow how we are doing it in other places. So just a small change in the import :)

mikehardy replied:
Ah, I didn't catch the file move, that makes sense

mikehardy commented on packages/analytics/lib/modular/index.d.ts:
```diff
@@ -0,0 +1,1188 @@
+import AddPaymentInfoEventParameters = FirebaseAnalyticsTypes.AddPaymentInfoEventParameters;
+import AddShippingInfoParameters = FirebaseAnalyticsTypes.AddShippingInfoParameters;
+import AddToCartEventParameters = FirebaseAnalyticsTypes.AddToCartEventParameters;
+import AddToWishlistEventParameters = FirebaseAnalyticsTypes.AddToWishlistEventParameters;
+import Analytics = FirebaseAnalyticsTypes.Module;
+import AnalyticsCallOptions = FirebaseAnalyticsTypes.AnalyticsCallOptions;
+import AnalyticsSettings = FirebaseAnalyticsTypes.AnalyticsSettings;
+import BeginCheckoutEventParameters = FirebaseAnalyticsTypes.BeginCheckoutEventParameters;
+import CampaignDetailsEventParameters = FirebaseAnalyticsTypes.CampaignDetailsEventParameters;
+import CustomEventName = FirebaseAnalyticsTypes.CustomEventName;
+import EarnVirtualCurrencyEventParameters = FirebaseAnalyticsTypes.EarnVirtualCurrencyEventParameters;
+import EventParams = FirebaseAnalyticsTypes.EventParams;
+import FirebaseApp = ReactNativeFirebase.FirebaseApp;
+import GenerateLeadEventParameters = FirebaseAnalyticsTypes.GenerateLeadEventParameters;
+import JoinGroupEventParameters = FirebaseAnalyticsTypes.JoinGroupEventParameters;
+import LevelEndEventParameters = FirebaseAnalyticsTypes.LevelEndEventParameters;
+import LevelStartEventParameters = FirebaseAnalyticsTypes.LevelStartEventParameters;
+import LevelUpEventParameters = FirebaseAnalyticsTypes.LevelUpEventParameters;
+import LoginEventParameters = FirebaseAnalyticsTypes.LoginEventParameters;
+import PostScoreEventParameters = FirebaseAnalyticsTypes.PostScoreEventParameters;
+import PurchaseEventParameters = FirebaseAnalyticsTypes.PurchaseEventParameters;
+import RefundEventParameters = FirebaseAnalyticsTypes.RefundEventParameters;
+import RemoveFromCartEventParameters = FirebaseAnalyticsTypes.RemoveFromCartEventParameters;
+import ScreenViewParameters = FirebaseAnalyticsTypes.ScreenViewParameters;
+import SearchEventParameters = FirebaseAnalyticsTypes.SearchEventParameters;
+import SelectContentEventParameters = FirebaseAnalyticsTypes.SelectContentEventParameters;
+import SelectItemEventParameters = FirebaseAnalyticsTypes.SelectItemEventParameters;
+import SelectPromotionEventParameters = FirebaseAnalyticsTypes.SelectPromotionEventParameters;
+import SetCheckoutOptionEventParameters = FirebaseAnalyticsTypes.SetCheckoutOptionEventParameters;
+import ShareEventParameters = FirebaseAnalyticsTypes.ShareEventParameters;
+import SignUpEventParameters = FirebaseAnalyticsTypes.SignUpEventParameters;
+import SpendVirtualCurrencyEventParameters = FirebaseAnalyticsTypes.SpendVirtualCurrencyEventParameters;
+import UnlockAchievementEventParameters = FirebaseAnalyticsTypes.UnlockAchievementEventParameters;
+import ViewCartEventParameters = FirebaseAnalyticsTypes.ViewCartEventParameters;
+import ViewItemEventParameters = FirebaseAnalyticsTypes.ViewItemEventParameters;
+import ViewItemListEventParameters = FirebaseAnalyticsTypes.ViewItemListEventParameters;
+import ViewPromotionEventParameters = FirebaseAnalyticsTypes.ViewPromotionEventParameters;
+import ViewSearchResultsParameters = FirebaseAnalyticsTypes.ViewSearchResultsParameters;
+import ConsentSettings = FirebaseAnalyticsTypes.ConsentSettings;
+import SettingsOptions = FirebaseAnalyticsTypes.SettingsOptions;
```
I'm not sure importing the sub-types into this namespace adds value - it appears that one of them (AnalyticsCallOptions) is used pretty frequently but the rest are used only once for the most part.

So there is a huge block of types up at the top here (which adds to the mental load for the developer + reviewer to remember the local types) without the benefit of really eliminating much extra text (the `FirebaseAnalyticsTypes.` prefix that is eliminated would only show up once anyway...)

I'd prefer fewer new types (created in this namespace by the named imports) and would rather see the types directly referenced as `FirebaseAnalyticsTypes.<subtype>` in the referencing code below

mikehardy reviewed (commented):
Mostly comments + questions, and one stated preference

mikehardy committed:
fix(app, sdk): adopt firebase-ios-sdk 10.11.0

the underlying SDK exposes some new features but
they are not exposed here, but we do get the benefit
of the underlying bugfixes

exaby73 committed:
revert modular index.js changes with whitespace

exaby73 committed:
revert modular index.js changes with whitespace

exaby73 committed:
remove single imports in index.d.ts

mikehardy committed:
fix(remote-config, ios): workaround firebase-ios-sdk#11458 until SDK v10.12.0

the impact of the workaround is that on iOS, if you subscribe to remote-config
realtime updates at all, the web socket to listen for updates will be opened and
never closed, even if you unsubscribe all listeners

this should be reverted when 10.12.0 is adopted

mikehardy committed:
test(remote-config, ios): workaround for firebase-ios-sdk#11462

skip tests when the error condition indicated by the upstream issue is encountered
anticipate this may be reverted with release of firebase-ios-sdk 10.12.0

Salakar committed:
chore(release): release packages

mikehardy commented on packages/analytics/lib/index.d.ts:
```diff
@@ -229,6 +229,7 @@ export namespace FirebaseAnalyticsTypes {
      */
     value: number;
   }
+
```
there are still spacing changes all over, expanding the diff for no functional purpose 🤔

exaby73 replied:
Unfortunately, it's not trivial to remove these spaces that prettier added without going line by line over hundreds of lines of code over the entire PR. If you have any ideas on how to achieve this without me having to spend literal hours removing spaces, let me know :)

mikehardy reviewed (changes requested):
still some tangential spacing changes mixed with the diff, still hoping for a link to authoritative reference for string constants

mikehardy committed:
docs(auth): fix docs render issue in AuthSettings

unterminated tag was breaking Vercel docs build, I should have caught that on
the PR review, it was obvious in hindsight

MayoudP committed:
docs(perf): Update wrong typo for startScreenTrace

richerfu committed:
docs(contributing): fix promise typo

brianpeiris committed:
docs(dynamic-links): Fix link to firebase docs (#7186)

mikehardy committed:
build(deps): update all updatable javascript packages

- lerna has a semver major to absorb
- ESM-only packages left alone (node-fetch, inquirer)
- superstruct still quite out of date
- detox requires switching to jest from mocha
- react-native + related packages will be separate
- expo will be separate

mikehardy committed:
test(app): `localhost` is `::1` on some hosts, `127.0.0.1` is required

without this, hosts with IPv6 resolve localhost to ::1 and the firebase
emulator isn't listening there, so our helpers cannot connect

mikehardy committed:
test(auth): convert users delete / disable helpers to node-fetch style

something about the http module and the previous style of posting to the
emulator wasn't working, in the same way remote-config realtime template
update cloud function wasn't working even though it looked perfect

node-fetch works great though, so convert the two helpers to that

this may be related to a node v18 switch pending / in testing locally where
http module works for v16 but not v18, but node-fetch works for both

mikehardy committed:
build(deps): bump from JDK11 to JDK17

required one tweak to gradle.properties to remove argument dropped in JDK17

mikehardy committed:
build(deps): bump node to v18

note that there is a specific issue with node >= v18.16 where
jest will throw 'TypeError: Cannot redefine property: performance'
so for jest tests we pin it to that specific version

everyone else can just use the generic v18

mikehardy committed:
build(deps): react-native 0.72

mikehardy committed:
test(firestore): skip batch write limit test that fails now

apparently > 500 batched writes will pass the firestore emulator now,
but official documentation indicates 500 is still the limit.

Should raise with upstream

mikehardy committed:
fix(app, ios): incorporate firebase-ios-sdk 10.12.0 (#7231)

* fix(app, ios): incorporate firebase-ios-sdk 10.12.0

* fix(remote-config, ios): remove workarounds for realtime updates on iOS

with the new firebase-ios-sdk 10.12.0 all the upstream fixes required have
been incorporated + released

mikehardy commented:
Appears to have taken conflicts as I added `getSessionId` support from #6633

Salakar committed:
chore(release): release packages

donaldkwong committed:
feat(auth, revokeToken): sign in with apple revokeToken API (#7239)

* feat(revoke): add support for revokeToken API

* Fixed there/their

* style(auth, lint): result of `yarn lint:android && yarn lint:ios:fix`

---------

Co-authored-by: Mike Hardy <github@mikehardy.net>

Salakar committed:
chore(release): release packages

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/react-native-firebase?pullRequest=7109) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/invertase/react-native-firebase?pullRequest=7109) before we can accept your contribution.<br/>**16** out of **18** committers have signed the CLA.<br/><br/>:white_check_mark: exaby73<br/>:white_check_mark: skam22<br/>:white_check_mark: Bilal-Abdeen<br/>:white_check_mark: zhiqingchen<br/>:white_check_mark: mikehardy<br/>:white_check_mark: aadito123<br/>:white_check_mark: galezra<br/>:white_check_mark: ibobo<br/>:white_check_mark: falleco<br/>:white_check_mark: MayoudP<br/>:white_check_mark: brianpeiris<br/>:white_check_mark: richerfu<br/>:white_check_mark: billykwok<br/>:white_check_mark: donaldkwong<br/>:white_check_mark: dlackty<br/>:white_check_mark: ashuvssut<br/>:x: Salakar<br/>:x: dependabot[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/react-native-firebase?pullRequest=7109) it.</sub>

exaby73 committed:
Merge branch 'main' into feat/analytics-modular-types

exaby73 committed:
fix duplicated function caused by merge

mikehardy commented on packages/analytics/lib/modular/index.js:
```diff
@@ -3,7 +3,7 @@ import { firebase } from '..';
  * Returns a Analytics instance for the given app.
  * @param app - FirebaseApp. Optional.
- * @returns {Analytics}
+ * @returns {import('..').FirebaseAnalyticsTypes.Module}
```
Is this true? This does not seem like a valid change

exaby73 replied:
`Analytics` is not recognised. The linter gives the following error: `Cannot find name 'Analytics'.`. In the `d.ts` files, `Analytics` is an alias of `FirebaseAnalyticsTypes.Module`. Hence why I added this. The types work as expected

mikehardy replied:
resolved with typedefs per Russell's suggestion

mikehardy commented on packages/analytics/lib/modular/index.js:
```diff
@@ -17,7 +17,7 @@ export function getAnalytics(app) {
  * @param app - FirebaseApp.
  * @param options - `AnalyticsSettings`. Web only.
- * @returns {Analytics}
+ * @returns {import('..').FirebaseAnalyticsTypes.Module}
```
This does not seem like a valid change

mikehardy reviewed (changes requested):
This is still not ready to go - I saw a couple things I did not see before and there was an unanswered question

The merge commits make the commit change indecipherable in general and the spacing changes make the combined diff difficult to look at but I reviewed it again nonetheless

Please check the new comments and we can get this in

exaby73 committed:
add reference to EventNameString

exaby73 committed:
revert change to setContent params

exaby73 commented on packages/analytics/lib/modular/index.js:
```diff
@@ -577,7 +628,7 @@ export function isSupported() {
  */
 // eslint-disable-next-line
-export function setConsent(consentSettings) {
+export function setConsent(analytics, consentSettings) {
```
https://github.com/invertase/react-native-firebase/pull/7109/commits/1c8255ecb5a3f13ddef18497fc92b0fe12a86d88

mikehardy committed:
fix(types): use typedef for clean types in jsdoc

mikehardy reviewed (dismissed):
typedefs are used to get clean types in the jsdoc now, Nabeel got the reference to the upstream event names in there, just pending CI

mikehardy reviewed (approved):
CI went green but I had a copy/paste error in last push, fixed re-pushed and good to go
